rfc/system/5598-deterministic-errors-distributed-training.md

```diff
@@ -0,0 +1,132 @@
+
+
+Furthermore, we propose to add an optional `timestamp` attributes to all [flytekit exceptions](https://github.com/flyteorg/flytekit/tree/master/flytekit/exceptions).
+
+The flytekit pytorch elastic plugin, for instance, catches `ChildFailedError`s [here](https://github.com/flyteorg/flytekit/blob/77d056ab9fda40ec6b2312a4d197b9107cdb70dc/plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py#L449), would extract the timestamp, and re-raise it as a Flyte exception which contains a timestamp. (Other plugins, e.g. non-elastic pytorch, which don't come with built-in exception types that include error timestamps, can themselves record the timestamp when the `task_function` raises an exception.)
+
+The entrypoint `pyflyte-execute` will transfer the timestamp from the flytekit exception into the protobuf `ContainerError`. It will also set the `worker` attribute of the `ContainerError` according to the `FLYTE_INTERNAL_WORKER_NAME` environment variable introduced above.
+
+### Flytepropeller/Flyteplugins - Aggregate the errors in the backend
+
+In the [kubernetes plugin machinery](https://github.com/flyteorg/flyte/blob/815f85d0ce90a3ace61cce17c0bfb441ac2dbcc3/flyteplugins/go/tasks/pluginmachinery/k8s/plugin.go) we propose to define the error aggregation strategy and allow plugins to configure it via their `PluginProperties`:
+
+```go
+type ErrorAggregationStrategy int
+
+const (
+	// Single error file from a single container
+	Default ErrorAggregationStrategy = iota
+
+	// Earliest error from potentially multiple error files
+	Earliest
+)
+
+// System level properties that this Plugin supports
+type PluginProperties struct {
+	...
+	ErrorAggregationStrategy ErrorAggregationStrategy
+}
+```
+
+Currently, [here](https://github.com/flyteorg/flyte/blob/4514860cf56ba62717f6c207f269410a8c1a5461/flytepropeller/pkg/controller/nodes/task/k8s/plugin_manager.go#L290) in the plugin manager, upon completion of a node execution, a new [`RemoteFileOutputReader`](https://github.com/flyteorg/flyte/blob/d6da838627d57cd27d60beea004e974ce1fb3ca5/flyteplugins/go/tasks/pluginmachinery/ioutils/remote_file_output_reader.go#L14) is constructed which is responsible for reading the error file uploaded to blob storage. This `RemoteFileOutputReader` implements the [`OutputReader` interface](https://github.com/flyteorg/flyte/blob/1e54d21c4d4ee74245f799a57b4bb8a5534e8368/flyteplugins/go/tasks/pluginmachinery/io/iface.go#L32).
+
+We propose to implement a new `MultiErrorFileRemoteFileOutputReader` which (for future flexibility) can be configured with the different strategies we define. Initially, the only available strategy will be `"earliest"` which the RFC authors aim to use for the kubeflow pytorch plugin. This output reader will search for all error files in the `/errors` folder under the raw output prefix and aggregate the error as specified by the strategy.
```

Comment:
Instead of a `MultiErrorFileRemoteFileOutputReader` class, why not have `RemoteFileOutputReader` take an `ErrorAggregationStrategy`?

Reply:
The `RemoteFileOutputReader`'s main goal is to read outputs. We can have multiple classes to extract the errors, which would be an internal detail to `RemoteFileOutputReader`.

Reply:
The `RemoteFileOutputReader`'s main goal is to read outputs. We can have multiple classes to extract the errors, which would be an internal detail to `RemoteFileOutputReader`.